### DISCUSSION TITLE
Keybinding to focus to a window

### DISCUSSION THREAD

**Discussion by mohammedbilalns** [YaLTeR/niri#467] [Ideas]
Can I configure a keybinding to focus on a window, such as Mod+Shift+n, to focus on the nth window from the left

**Comment by YaLTeR** [comment]
There's no such functionality at the moment

**Comment by Alanon202** [comment]
@YaLTeR Sorry to necro such an old discussion, but it’s closest to what I’m looking into. 

I’m thinking on how to set up Niri so that the same keybinding is able to 1) start an app, or, 2) focus on that app if the app is already running, wherever it is. Some apps do this automatically, but others do not, so it can be very beneficial and speed up the workflow. I wasn’t able to find anything about this in the wiki, so I presume it’s not a feature at the moment?

I was thinking of creating a wrapper script that will do the heavy lifting of checking if a process is active and launching it, by passing the app name as a parameter. However, I would need to hook into something like niri ipc to be able to focus the window? Is this possible?

**Reply by YaLTeR** [reply]
You can script it with `niri msg windows` and `niri msg action focus-window`

**Reply by Brisingr05** [reply]
> 1) start an app, or, 2) focus on that app if the app is already running, wherever it is

For this, you can use [nirius](https://git.sr.ht/~tsdh/nirius). Example:
```kdl
binds {
    Mod+W {
        spawn "nirius" "focus-or-spawn" "--app-id" "Mullvad Browser" "--" "mullvad-browser"
    }
}
```
If a window with app-id matching "Mullvad Browser" already exists, it will be focused. Otherwise it will be launched using the mullvad-browser command.

**Reply by Alanon202** [reply]
@Brisingr05 Thanks! I wish I’d seen this sooner as I hacked something together in bash yesterday, [here it is](https://github.com/Alanon202/nirirun).

Probably a lot less elegant than nirius, but has some capabilities it doesn’t at the moment. I wish I could make my script work for pinning picture-in-picture, but that’ too much to ask from bash...

**Reply by Brisingr05** [reply]
> pinning picture-in-picture

That's what `toggle-follow-mode` is for in `nirius`. It will make a window "follow" you when switching workspaces:

https://github.com/user-attachments/assets/ea114e0b-9b2c-470e-8cb6-deef2c8a1083

**Reply by Fireye04** [reply]
> > pinning picture-in-picture
> 
> That's what `toggle-follow-mode` is for in `nirius`. It will make a window "follow" you when switching workspaces:
> 
> https://github.com/user-attachments/assets/ea114e0b-9b2c-470e-8cb6-deef2c8a1083

Unfortunately it doesn't work in windowrules yet. I have it binded to Mod+P though which works great for now though :P

**Reply by Brisingr05** [reply]
> Unfortunately it doesn't work in windowrules yet. I have it binded to Mod+P though which works great for now though :P

Ah right. Relevant GitHub Discussion by the `nirius` author: https://github.com/YaLTeR/niri/discussions/1945

**Reply by Alanon202** [reply]
Ah ok. So you’d have to start the ```toggle-follow-mouse``` sometime _after_ PiP has started playing for it to work? What happens if you exit a PiP window, then respawn it? 

For instance, I have it set up so that a PiP window spawns as soon as I unfocus a playing tab. But if I return to that tab the PiP goes away, also if playback is paused for too long, the tab hibernates, killing the PiP window. Would all of these cases require a re-application of the command?

This did give me an idea to make a simple title match in bash that will just move the PiP to where the user is.
